### PR TITLE
[feat] Hybrid model ep overlapping main

### PR DESCRIPTION
**Pull request by Wohox**
# What does this PR do ?
## Summary

This MR adds grouped HybridStack support and extends EP-overlap scheduling/checkpoint compatibility to grouped hybrid layers.

Key changes:
- Add bracketed HybridStack group syntax, e.g. `[*-]`, `[*E]`, `M[M*]-`.
- Build bracketed groups as nested `HybridStack` instances.
- Reject invalid recursion such as nested brackets inside a group.
- Keep MoE constrained to the last symbol inside a group for overlap scheduling.
- Rename/migrate `TransformerSchedulePlan` usage to `HybridStackSchedulePlan`, preserving backward-compatible aliases.
- Add HybridStack fine-grained callables so grouped hybrid layers can participate in combined 1F1B / EP overlap.
- Make grouped HybridStack checkpoints use Transformer-compatible logical layer keys.
- Make `HybridModel.sharded_state_dict()` match GPT behavior by dropping empty `output_layer._extra_state`.

## Performance Improvement Parity
With TP1CP1PP4VPP2EP64 on 16 layer DSv3 test, hybird model and gpt model achieve performance improvement with EP overlap and achieves throughput parity. [w&b](https://wandb.ai/megatron-core-moe-dev/pingtianl-deepseek-v3-gpt-full-attention-H100-benchmark?nw=nwuserpingtianl&panelDisplayName=throughput&panelSectionName=Charts&xAxisMin=1&xAxisMax=80.59717830724469)
Screenshot 2026-06-01 at 19 23 39

## Checkpoint Compatibility

Grouped HybridStack checkpoint keys now canonicalize to logical Transformer layer keys. For example, `[*-]` maps both attention and MLP weights to:

```text
decoder.layers.N.self_attention.*
decoder.layers.N.mlp.*
```

instead of using separate physical layer ids. `final_norm` is also sharded as `final_layernorm` for Transformer compatibility.

Verified both directions:
- Transformer checkpoint -> HybridModel: load succeeded, one train step completed, checkpoint saved, validation/test completed.
- HybridModel checkpoint -> Transformer: load succeeded, one train step completed, checkpoint saved, validation/test completed.

Cross-load status:

```text
STATUS save_final_hybrid 0
STATUS transformer_ckpt_to_hybrid 0
STATUS final_hybrid_ckpt_to_transformer 0
```

## Validation

unit tests:

```bash
python -m pytest -q \
  tests/unit_tests/ssm/test_hybrid_layer_allocation.py \
  tests/unit_tests/ssm/test_hybrid_block.py \
  tests/unit_tests/models/test_hybrid_model.py::TestHybridModel::test_grouped_sharded_state_dict_uses_transformer_checkpoint_keys
```

Result:

```text
86 passed, 63 warnings
```

Also ran:
- `git diff --check`
- DeepSeek-V3 deterministic Transformer/Hybrid baseline and EP-overlap smoke tests
- Transformer VPP4 baseline/overlap comparison
- strict model-key checkpoint cross-load checks using `--dist-ckpt-strictness raise_unexpected`

## Notes

The checkpoint cross-load validation intentionally uses model-weight / finetune-style loading with optimizer and RNG state skipped. Full non-finetune resume across Transformer vs grouped HybridModel is still expected to hit config arg mismatches because Transformer uses logical `num_layers=16`, while grouped Hybrid uses physical symbols grouped into 16 logical layers.
## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Comment by Connor-XY** [comment]
Pushed five commits from the split branches onto this branch so it doesn't drift from what will actually merge:

- `fix(hybrid): preserve MTP postprocess state`
- `Fix grouped hybrid review issues`
- `Fix grouped hybrid unit test failures`
- `Address review: scope hybrid checkpoint key rename, honor output_processor`
- `Opt directly-constructed grouped HybridStack into transformer keys`

The last two answer the strict review on #4942. The one that matters here: `test_grouped_sharded_state_dict_uses_transformer_checkpoint_keys` was failing on this branch because it asserts `decoder.final_layernorm.weight` but `_sharded_state_dict` never renamed the key. That rename now exists and is gated on `transformer_sharded_keys`, which `HybridModel` sets when the layer pattern has a bracketed group — so grouped models cross-load with GPT and everything else keeps `decoder.final_norm.*`, matching what `tools/checkpoint/gpt_hybrid_conversion.py` writes.

All fast-forward, nothing rewritten. @Wohox shout if you'd rather I back any of it out.
